spark/src/main/scala/org/apache/iceberg/spark/SparkTableUtil.scala

```diff
@@ -27,86 +27,153 @@ import org.apache.hadoop.fs.{Path, PathFilter}
+  }
 
-    val partitions: Seq[(Map[String, String], Option[String], Option[String])] =
-      Hive.partitions(spark, table).map { p: CatalogTablePartition =>
-        (p.spec, p.storage.locationUri.map(String.valueOf(_)), p.storage.serde)
-      }
+  /**
+   * Returns all partitions in the table.
+   *
+   * @param spark a Spark session
+   * @param tableIdent a table identifier
+   * @return all table's partitions
+   */
+  def getPartitions(spark: SparkSession, tableIdent: TableIdentifier): Seq[SparkPartition] = {
+    val catalog = spark.sessionState.catalog
+    val catalogTable = catalog.getTableMetadata(tableIdent)
 
-    partitions.toDF("partition", "uri", "format")
+    catalog
+      .listPartitions(tableIdent)
+      .map(catalogPartition => toSparkPartition(catalogPartition, catalogTable))
   }
 
   /**
-    * Returns a DataFrame with a row for each partition that matches the specified 'expression'.
-    *
-    * @param spark a Spark session.
-    * @param table name of the table.
-    * @param expression The expression whose matching partitions are returned.
-    * @return a DataFrame of the table partitions.
-    */
-  def partitionDFByFilter(spark: SparkSession, table: String, expression: String): DataFrame = {
-    import spark.implicits._
+   * Returns partitions that match the specified 'predicate'.
+   *
+   * @param spark a Spark session
+   * @param table a table name and (optional) database
+   * @param predicate a predicate on partition columns
+   * @return matching table's partitions
+   */
+  def getPartitionsByFilter(spark: SparkSession, table: String, predicate: String): Seq[SparkPartition] = {
+    val tableIdentifier = spark.sessionState.sqlParser.parseTableIdentifier(table)
+
+    val unresolvedPredicateExpr = spark.sessionState.sqlParser.parseExpression(predicate)
+    val resolver = spark.sessionState.analyzer.resolver
+    val plan = spark.table(table).queryExecution.analyzed
+    val resolvedPredicateExpr = unresolvedPredicateExpr.transform {
+      case attr: UnresolvedAttribute =>
+        plan.resolve(attr.nameParts, resolver) match {
+          case Some(resolvedAttr) => resolvedAttr
+          case None => throw new IllegalArgumentException(s"Could not resolve $attr using columns: ${plan.output}")
+        }
+    }
 
-    val expr = spark.sessionState.sqlParser.parseExpression(expression)
-    val partitions: Seq[(Map[String, String], Option[String], Option[String])] =
-      Hive.partitionsByFilter(spark, table, expr).map { p: CatalogTablePartition =>
-        (p.spec, p.storage.locationUri.map(String.valueOf(_)), p.storage.serde)
-      }
+    getPartitionsByFilter(spark, tableIdentifier, resolvedPredicateExpr)
+  }
+
+  /**
+   * Returns partitions that match the specified 'predicate'.
+   *
+   * @param spark a Spark session
+   * @param tableIdent a table identifier
+   * @param predicate a predicate on partition columns
+   * @return matching table's partitions
+   */
+  def getPartitionsByFilter(
```

Comment:
Do we have a test for this?

Reply:
Yes, we do test this in `TestSparkTableUtil` as well as in `TestSparkTableUtilWithInMemoryCatalog`.

One example is [here](https://github.com/apache/incubator-iceberg/pull/624/files#diff-feacbac75264b25b7ec43e7b99d250a4R133).